File: frontend/lib/util/goban_hash.dart
import 'dart:math';

import 'package:frontend/models/joseki.dart';

// const int mod = (1 << 61) - 1;
// const int mask30 = (1 << 30) - 1;
// const int mask31 = (1 << 31) - 1;
// const int baseRow = 37;
// const int baseCol = 23;
// final List<int> row = generateArray(baseRow);
// final List<int> col = generateArray(baseCol);

// int mul(int a, int b) {
//   int au = a >> 31;
//   int ad = a & mask31;
//   int bu = b >> 31;
//   int bd = b & mask31;
//   int mid = ad * bu + au * bd;
//   int midu = mid >> 30;
//   int midd = mid & mask30;
//   return calcMod(au * bu * 2 + midu + (midd << 31) + ad * bd);
// }

// int calcMod(int x) {
//   int xu = x >> 61;
//   int xd = x & mod;
//   int res = xu + xd;
//   if (res >= mod) res -= mod;
//   return res;
// }

// List<int> generateArray(int base) {
//   List<int> array = [1];
//   for (int i = 1; i < boardSize; i++) {
//     array.add(mul(array.last, base));
//   }
//   return array;
// }

// // rolling hash
// int gobanHash(StoneMatrix stoneMatrix) {
//   int hash = 0;
//   for (int i = 0; i < boardSize; i++) {
//     for (int j = 0; j < boardSize; j++) {
//       if (stoneMatrix[i][j].color == StoneColor.empty) continue;
//       int val = mul(row[i], col[j]);
//       if (stoneMatrix[i][j].color == StoneColor.black) {
//         hash += val;
//       } else {
//         hash -= val;
//       }
//       hash = calcMod(hash);
//     }
//   }
//   return hash;
// }

final List<List<List<int>>> hashTable = generateTable();

List<List<List<int>>> generateTable() {
  Random random = Random(0);
  List<List<List<int>>> table = [];
  for (int i = 0; i < boardSize; i++) {
    List<List<int>> row = [];
    for (int j = 0; j < boardSize; j++) {
      List<int> col = [];
      for (int k = 0; k < 2; k++) {
        col.add(random.nextInt(1 << 30) << 30 | random.nextInt(1 << 30));
      }
      row.add(col);
    }
    table.add(row);
  }
  return table;
}

int gobanHash(StoneList captured, Stone stone, int originalHash) {
  int hash = originalHash;
  if (stone.color != StoneColor.empty) {
    hash ^= hashTable[stone.x][stone.y][stone.color.index];
  }
  for (int i = 0; i < captured.length; i++) {
    Stone cap = captured[i];
    if (cap.color != StoneColor.empty) {
      hash ^= hashTable[cap.x][cap.y][cap.color.index];
    }
  }
  return hash;
}

File: web/src/utils/gobanHash.ts
import { boardSize, type Stone } from "../models/model"

class Random {
  x: number
  y: number
  z: number
  w: number
  constructor(seed = 88675123) {
    this.x = 123456789
    this.y = 362436069
    this.z = 521288629
    this.w = seed
  }

  next(): number {
    let t = this.x ^ (this.x << 11)
    this.x = this.y
    this.y = this.z
    this.z = this.w
    return (this.w = this.w ^ (this.w >>> 19) ^ (t ^ (t >>> 8)))
  }

  nextInt(min: number, max: number): number {
    const r = Math.abs(this.next())
    return min + (r % (max + 1 - min))
  }
}

const generateTable = (): Array<Array<Array<number>>> => {
  const random = new Random()
  const table: Array<Array<Array<number>>> = []
  for (let i = 0; i < boardSize; i++) {
    const row: Array<Array<number>> = []
    for (let j = 0; j < boardSize; j++) {
      const col: Array<number> = []
      for (let k = 0; k < 2; k++) {
        col.push(
          (random.nextInt(0, 1 << 30) << 30) | random.nextInt(0, 1 << 30)
        )
      }
      row.push(col)
    }
    table.push(row)
  }
  return table
}

export const gobanHash = (
  captured: Array<Stone>,
  stone: Stone,
  originalHash: number
) => {
  const hashTable = generateTable()
  let hash = originalHash
  if (stone.color != "empty") {
    hash ^= hashTable[stone.x][stone.y][stone.color == "black" ? 0 : 1]
  }
  for (let i = 0; i < captured.length; i++) {
    const cap = captured[i]
    if (cap.color != "empty") {
      hash ^= hashTable[cap.x][cap.y][cap.color == "black" ? 0 : 1]
    }
  }
  return hash
}
